image_writers_guide/sti.adoc

```diff
@@ -5,8 +5,142 @@
 :icons:
 :experimental:
 :toc: macro
-:toc-title: 
+:toc-title:
 
 toc::[]
 
-== Overview
\ No newline at end of file
+== Overview
+
+The main advantage of using sti for building reproducible docker image ease of use
+for developer. To meet that criteria you, as a builder image author, should be aware
+of the two basic requirements for the best possible sti performance, these are:
+
+* link:[#Required image contents]
+* link:[#STI scripts]
+
+
+== Required image contents
+
+The build process consists of three fundamental elements which are combined into
+final docker image, the three are: sources, scripts and builder image. During the
+build process sti must place sources and scripts inside that builder image. To do
+so sti creates a tar file containing the two and then streams that file into the
+builder image. Before executing `assemble` script, sti untars that file and places
+its contents into the location specified with `--location` flag or `STI_LOCATION`
+environment variable from the builder image (default location is `/tmp`). For this
+to happen your image must supply tar archiving utility (command `tar` available in `$PATH`)
+and command line interpreter (command `/bin/sh`). Doing so will allow your image to
+use the fastest possible build path, because in all other cases when either
+`tar` or `/bin/sh` is not available, sti build will be forced to perform an additional
+docker build to put both sources and scripts inside the image and only then run the
+usual sti build procedure (sti will do this automatically if tar and /bin/sh are not found).
+See the following diagram for how build workflow looks like:
+
+image:sti-flow.png[STI workflow]
+
+\* Run build's responsibility is to untar the sources, scripts and artifacts (if such
+exist) and invoke `assemble` script. If this is second run (after catching `tar`/`/bin/sh`
+error) it's responsible only for invoking `assemble` script, since both scripts and
+sources are already there.
+
+
+== STI scripts
+
+STI expects you, as the builder image author, to supply the following scripts:
+
+* required:
+    * link:[#assemble]
+    * link:[#run]
+* optional:
+    * link:[#save-artifacts]
+    * link:[#usage]
+
+All of the scripts can be written in any programming language, as long as it is
+executable inside the builder image.
+
+=== assemble
+
+The `assemble` script is responsible for building the application artifacts from source,
+and place them into appropriate directories inside the image. The workflow for `assemble` is:
+
+1. Restore build artifacts (in case you want to support incremental builds, make sure
+   to define [save-artifacts](#save-artifacts)) as well.
+1. Place the application source in desired location.
```

Comment:
it's handy when you need to add new steps in the middle (no need to renumber) but it's horrible if you're trying to refer to steps from other steps (since you can't see what number the steps are).

as long as you're not doing that, it's probably fine.